common/src/main/java/com/gregtechceu/gtceu/client/model/PipeModel.java

```diff
@@ -174,11 +176,16 @@ public TextureAtlasSprite getParticleTexture() {
         return sideSprite;
     }
 
+    private final Function<Optional<Direction>, List<BakedQuad>> modelBaker = Util.memoize(direction ->
+            bakeQuads(direction.orElse(null), ITEM_CONNECTIONS)
+    );
+
     @Environment(EnvType.CLIENT)
     public void renderItem(ItemStack stack, ItemDisplayContext transformType, boolean leftHand, PoseStack matrixStack, MultiBufferSource buffer, int combinedLight, int combinedOverlay, BakedModel model) {
         IItemRendererProvider.disabled.set(true);
         Minecraft.getInstance().getItemRenderer().render(stack, transformType, leftHand, matrixStack, buffer, combinedLight, combinedOverlay,
-                (ItemBakedModel) (state, direction, random) -> bakeQuads(direction, ITEM_CONNECTIONS));
+                (ItemBakedModel) (state, direction, random) -> modelBaker.apply(Optional.ofNullable(direction))
```

Comment:
add cache for `bakeQuads` instead, I removed it before, but it should be fine rn. and dont forget to clear it when the client resource is reloading. 
Besides, make sure the thread safety of the cache, baking will perform on asynch threads in fact.

Reply:
Ah yeah that's probably a good idea, I'll add it. Minecraft's memoize function is threadsafe by default iirc.

Reply:
I've tried this but the performance overhead is even worse if I'm using a record containing the direction and connections as a key.  
However, I've noticed that the models for placed blocks are cached either way.

I'm just gonna cache only the item rendering instead, but I'll add a way of invalidating the cache on resource reload.

Reply:
Just checked - we don't even need to invalidate caches on resource reload, since pipe models are not affected by resources.

Reply:
Wait nevermind, we probably do need to invalidate on reload, since the baked quads contain the textures if I understand it correctly.

Reply:
I just verified this and the textures reload fine, even if we're not invalidating the cache.